FD2: Add support for updating rt args in place

tt-aho opened the pull request.

tt-aho committed:
#0: Add support for updating rt args in place
Co-authored-by: Austin Ho <aho@tenstorrent.com>
Co-authored-by: Allan Liu <aliu@tenstorrent.com>

aliuTT commented on tt_metal/impl/dispatch/command_queue.cpp:
```diff
@@ -270,109 +270,230 @@ void EnqueueProgramCommand::assemble_preamble_commands() {
-                    break;
+// Generate command sequence for unique (unicast) and common (multicast) runtime args
+void EnqueueProgramCommand::assemble_runtime_args_commands() {
+    static const std::vector<uint32_t> processor_to_l1_arg_base_addr = {
```
can you leave a comment here that links to this:
```
enum class RISCV : uint8_t {
    BRISC = 0,
    NCRISC = 1,
    TRISC0 = 2,
    TRISC1 = 3,
    TRISC2 = 4,
    ERISC = 5,
    COMPUTE = 6,     // Encompasses TRISC0, TRISC1, and TRISC2
};
```

tt-aho replied:
Done

tt-aho commented:
@abhullar-tt I modified `add_dispatch_write_packed` `data_collection` to be a vector of pair of ptr + size. This change was for rt args since the length of each data vector was actually different and the padded size passed in is the max size, so we were technically reading from unknown memory for the smaller rt arg vectors. This hopefully shouldn't really affect perf for the other users of `add_dispatch_write_packed`, I modified them to add the length to the vector pairs, even though for these cases they could just use the passed in padded size.

mo-tenstorrent committed:
#0: Fix copy to ref
